dhctl/pkg/terraform/terraform_context.go

```diff
@@ -0,0 +1,347 @@
+// Copyright 2024 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terraform
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/config"
+	dstate "github.com/deckhouse/deckhouse/dhctl/pkg/state"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/util/tomb"
+)
+
+type TerraformContext struct {
+	terraformRunnerByName    map[string]RunnerInterface
+	terraformRunnerByNameMux sync.Mutex
+}
+
+func NewTerraformContext() *TerraformContext {
+	return &TerraformContext{
+		terraformRunnerByName: make(map[string]RunnerInterface),
+	}
+}
+
+func (f *TerraformContext) getOrCreateRunner(name string, createFunc func() RunnerInterface) RunnerInterface {
+	f.terraformRunnerByNameMux.Lock()
+	defer f.terraformRunnerByNameMux.Unlock()
+
+	// FIXME
+	//if r, hasKey := f.terraformRunnerByName[name]; hasKey {
+	//	return r
+	//}
+
+	r := createFunc()
+	f.terraformRunnerByName[name] = r
+	return r
+}
+
+// TODO(dhctl-for-commander): Use same tf-runner for check & converge in commander mode only, keep things as-is without changes
+func (f *TerraformContext) GetCheckBaseInfraRunner(metaConfig *config.MetaConfig, opts BaseInfraRunnerOptions) RunnerInterface {
+	return f.getOrCreateRunner(
+		fmt.Sprintf("check.base-infrastructure.%s.%s.%s", metaConfig.ProviderName, metaConfig.ClusterPrefix, metaConfig.Layout), func() RunnerInterface {
+			if opts.CommanderMode {
+				r := NewRunnerFromConfig(metaConfig, "base-infrastructure", opts.StateCache).
+					WithSkipChangesOnDeny(true).
+					WithVariables(metaConfig.MarshalConfig()).
+					WithAutoDismissDestructiveChanges(opts.AutoDismissDestructive).
+					WithAutoApprove(opts.AutoApprove)
+
+				r.WithAdditionalStateSaverDestination(opts.AdditionalStateSaverDestinations...)
+
+				tomb.RegisterOnShutdown("base-infrastructure", r.Stop)
+				return r
+			} else {
+				r := NewImmutableRunnerFromConfig(metaConfig, "base-infrastructure").
+					WithVariables(metaConfig.MarshalConfig()).
+					WithAutoApprove(true)
```

Comment:
Is it required for autoapprove == true, Looks like it allow potential destructive operation